pkg/podrestartmonitor/pod_restart_monitor.go

```diff
@@ -0,0 +1,183 @@
+
+const PodRestartMonitorName = "pod-restart-monitor"
+
+func init() {
+	problemdaemon.Register(
+		PodRestartMonitorName,
+		types.ProblemDaemonHandler{
+			CreateProblemDaemonOrDie: NewMonitorOrDie,
+			CmdOptionDescription:     "Set to config file paths."})
+}
+
+type podRestartMonitor struct {
+	configPath string
+	config     MonitorConfig
+	stop       chan struct{}
+	clientset  *kubernetes.Clientset
+	statuses   chan *types.Status
+}
+
+// NewMonitorOrDie create a new podRestartMonitor, panic if error occurs.
+func NewMonitorOrDie(configPath string) types.Monitor {
+	l := &podRestartMonitor{
+		configPath: configPath,
+	}
+
+	f, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		glog.Fatalf("Failed to read configuration file %q: %v", configPath, err)
+	}
+	err = json.Unmarshal(f, &l.config)
+	if err != nil {
+		glog.Fatalf("Failed to unmarshal configuration file %q: %v", configPath, err)
+	}
+
+	// Apply default configurations
+	if l.config.CheckInterval == 0 {
+		l.config.CheckInterval = Duration(5 * time.Minute)
+	}
+	if l.config.RestartThreshold == 0 {
+		l.config.RestartThreshold = 5
+	}
+	if l.config.Namespace == "" {
+		l.config.Namespace = "kube-system"
+	}
+	if l.config.PodSelector == "" {
+		l.config.PodSelector = "k8s-app=aws-node"
+	}
+
+	glog.Infof("Finish parsing pod restart monitor config file %s: %+v", l.configPath, l.config)
+	l.stop = make(chan struct{})
+
+	return l
+}
+
+func (l *podRestartMonitor) Start() (<-chan *types.Status, error) {
+	glog.Infof("Start pod restart monitor %s", l.configPath)
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	l.clientset, err = kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	l.statuses = make(chan *types.Status)
+
+	go l.monitorLoop()
+
+	return l.statuses, nil
+}
+
+func getNodeName() string {
+	return os.Getenv("NODE_NAME")
+}
+
+func (l *podRestartMonitor) Stop() {
+	glog.Infof("Stopping monitor %s", l.configPath)
+	close(l.stop)
+}
+
+const ConditionTooManyRestarts = "TooManyRestarts"
+
+// monitorLoop is the main loop of log monitor.
+func (l *podRestartMonitor) monitorLoop() {
+	nodeName := getNodeName()
+MainLoop:
+	for {
+		select {
+		case <-l.stop:
+			glog.Warningf("monitor stopped")
+			return
+		case <-time.After(time.Duration(l.config.CheckInterval)):
+			list, err := l.clientset.CoreV1().Pods(l.config.Namespace).List(metav1.ListOptions{
+				LabelSelector: l.config.PodSelector,
+				FieldSelector: "spec.nodeName=" + nodeName,
+			})
+			if err != nil {
+				glog.Errorf("failed to list pods: %+v", err)
+				continue
+			}
+			for _, pod := range list.Items {
+				for _, c := range pod.Status.ContainerStatuses {
+					count := c.RestartCount
+					if count > l.config.RestartThreshold {
+						l.statuses <- &types.Status{
+							Source: PodRestartMonitorName,
+							Events: nil,
+							Conditions: []types.Condition{
+								{Type: ConditionTooManyRestarts, Status: types.True, Transition: time.Now(),
```

Comment:
What happens when we want to monitor more than one pod (either by having a broad `PodSelector` or configuring multiple monitors)? It looks like the condition will collide. Is there a workaround to prevent this?

Reply:
I don't have a good answer because I don't have that use case. I suppose the condition type could be configurable.

Reply:
Looking at [official docs](https://kubernetes.io/docs/concepts/architecture/nodes/#condition), it appears that we can use reason for this. Let's keep `condition.Type` equal across all pods, but `condition.Reason` can contain the pod name to distinguish.